src/System.Text.Utf8/tests/RandomTests.cs

```diff
@@ -55,210 +56,107 @@ private static string GetStringLiteral(string s)
-                Assert.Equal(equal, s1FromSpan == s2);
-
-                Assert.Equal(equal, s1 == s2FromBytes);
-                Assert.Equal(equal, s1 == s2FromSpan);
-
-
-                Assert.Equal(equal, s1FromBytes.Equals(s2FromBytes));
-                Assert.Equal(equal, s1FromBytes.Equals(s2FromSpan));
-                Assert.Equal(equal, s1FromBytes.Equals(s2));
-
-                Assert.Equal(equal, s1FromSpan.Equals(s2FromBytes));
-                Assert.Equal(equal, s1FromSpan.Equals(s2FromSpan));
-                Assert.Equal(equal, s1FromSpan.Equals(s2));
-
-                Assert.Equal(equal, s1.EqualsUtf8String(s2FromBytes));
-                Assert.Equal(equal, s1.EqualsUtf8String(s2FromSpan));
-            }
+        public static object[][] LengthTestCases = {
+            new object[] { 0, default(Utf8String) },
+            new object[] { 0, Utf8String.Empty },
+            new object[] { 0, ""u8 },
+            new object[] { 4, "1258"u8 },
+            new object[] { 3, "\uABCD"u8 },
+            new object[] { 3, "\uABEE"u8 },
+            new object[] { 4, "a\uABEE"u8 },
+            new object[] { 5, "a\uABEEa"u8 },
+            new object[] { 8, "a\uABEE\uABCDa"u8 }
+        };
+        [Theory, MemberData("LengthTestCases")]
+        public void Length(int expectedLength, Utf8String s)
+        {
+            Assert.Equal(expectedLength, s.Length);
+        }
+
+        public static object[][] LengthInCodePointsTestCases = {
+            new object[] { 0, default(Utf8String) },
+            new object[] { 0, Utf8String.Empty },
+            new object[] { 0, ""u8 },
+            new object[] { 4, "1258"u8 },
+            new object[] { 1, "\uABCD"u8 },
+            new object[] { 1, "\uABEE"u8 },
+            new object[] { 2, "a\uABEE"u8 },
+            new object[] { 3, "a\uABEEa"u8 },
+            new object[] { 4, "a\uABEE\uABCDa"u8 }
+        };
+        [Theory, MemberData("LengthInCodePointsTestCases")]
+        public void LengthInCodePoints(int expectedLength, Utf8String s)
+        {
+            Assert.Equal(expectedLength, s.CodePoints.Count());
+        }
+
+        public static object[][] ToStringTestCases = {
+            new object[] { "", default(Utf8String) },
+            new object[] { "", Utf8String.Empty },
+            new object[] { "", ""u8 },
+            new object[] { "1258", "1258"u8 },
+            new object[] { "\uABCD", "\uABCD"u8 },
+            new object[] { "\uABEE", "\uABEE"u8 },
+            new object[] { "a\uABEE", "a\uABEE"u8 },
+            new object[] { "a\uABEEa", "a\uABEEa"u8 },
+            new object[] { "a\uABEE\uABCDa", "a\uABEE\uABCDa"u8 }
+        };
+        [Theory, MemberData("ToStringTestCases")]
+        public void ToString(string expected, Utf8String s)
+        {
+            Assert.Equal(expected, s.ToString());
+        }
+
+        public static object[][] StringEqualsTestCases = new object[][] {
+            // empty strings consistency
```

Comment:
These are a lot of test cases. Should we break them up to make it more readable? If you need to add comments to make sense of what you are doing, that's a smell that you can try and make things clearer in the code itself.